File: src/lib/utils/number-to-words.ts
/**
 * Number to Words Converter (Indian System)
 * 
 * Convert numbers to words for invoice amounts
 * Supports Indian numbering system (Lakhs, Crores)
 */

const ones = [
  '',
  'One',
  'Two',
  'Three',
  'Four',
  'Five',
  'Six',
  'Seven',
  'Eight',
  'Nine',
];

const teens = [
  'Ten',
  'Eleven',
  'Twelve',
  'Thirteen',
  'Fourteen',
  'Fifteen',
  'Sixteen',
  'Seventeen',
  'Eighteen',
  'Nineteen',
];

const tens = [
  '',
  '',
  'Twenty',
  'Thirty',
  'Forty',
  'Fifty',
  'Sixty',
  'Seventy',
  'Eighty',
  'Ninety',
];

/**
 * Convert a number less than 100 to words
 */
function convertTensToWords(num: number): string {
  if (num < 10) {
    return ones[num];
  } else if (num >= 10 && num < 20) {
    return teens[num - 10];
  } else {
    const ten = Math.floor(num / 10);
    const one = num % 10;
    return tens[ten] + (one > 0 ? ' ' + ones[one] : '');
  }
}

/**
 * Convert a number less than 1000 to words
 */
function convertHundredsToWords(num: number): string {
  if (num === 0) return '';

  const hundred = Math.floor(num / 100);
  const remainder = num % 100;

  let result = '';

  if (hundred > 0) {
    result += ones[hundred] + ' Hundred';
  }

  if (remainder > 0) {
    if (result !== '') {
      result += ' ';
    }
    result += convertTensToWords(remainder);
  }

  return result;
}

/**
 * Convert number to words (Indian system with Lakhs and Crores)
 */
export function numberToWordsIndian(num: number): string {
  if (num === 0) return 'Zero';

  // Handle negative numbers
  const isNegative = num < 0;
  num = Math.abs(num);

  // Split into integer and decimal parts
  const parts = num.toFixed(2).split('.');
  const integerPart = parseInt(parts[0]);
  const decimalPart = parseInt(parts[1]);

  let words = '';

  // Process integer part
  if (integerPart === 0) {
    words = 'Zero';
  } else {
    const crores = Math.floor(integerPart / 10000000);
    const lakhs = Math.floor((integerPart % 10000000) / 100000);
    const thousands = Math.floor((integerPart % 100000) / 1000);
    const hundreds = integerPart % 1000;

    if (crores > 0) {
      words += convertHundredsToWords(crores) + ' Crore ';
    }

    if (lakhs > 0) {
      words += convertHundredsToWords(lakhs) + ' Lakh ';
    }

    if (thousands > 0) {
      words += convertHundredsToWords(thousands) + ' Thousand ';
    }

    if (hundreds > 0) {
      words += convertHundredsToWords(hundreds);
    }

    words = words.trim();
  }

  // Add decimal part (paise)
  if (decimalPart > 0) {
    words += ' and ' + convertTensToWords(decimalPart) + ' Paise';
  }

  // Add "Rupees" and handle negative
  if (isNegative) {
    return 'Minus ' + words + ' Rupees Only';
  }

  return words + ' Rupees Only';
}

/**
 * Convert amount to words for invoice
 */
export function amountToWords(amount: number): string {
  return numberToWordsIndian(amount);
}

/**
 * Format number in Indian numbering system
 * Example: 1,00,00,000 (1 Crore)
 */
export function formatIndianNumber(num: number): string {
  const numStr = num.toString();
  const lastThree = numStr.substring(numStr.length - 3);
  const otherNumbers = numStr.substring(0, numStr.length - 3);

  if (otherNumbers !== '') {
    return (
      otherNumbers.replace(/\B(?=(\d{2})+(?!\d))/g, ',') + ',' + lastThree
    );
  } else {
    return lastThree;
  }
}

File: Backend/app/services/number_to_words.py
"""
Number to Words Converter (Indian Rupees)
Converts numeric amounts to words for invoices
"""

from decimal import Decimal


def amount_to_words(amount: Decimal) -> str:
    """
    Convert amount to words in Indian Rupees
    
    Args:
        amount: Decimal amount
        
    Returns:
        Amount in words (e.g., "One Thousand Two Hundred Thirty Four Rupees and Fifty Paise Only")
    """
    # Split into rupees and paise
    rupees = int(amount)
    paise = int((amount - rupees) * 100)
    
    result = []
    
    if rupees > 0:
        rupees_words = _number_to_words_indian(rupees)
        if rupees == 1:
            result.append(f"{rupees_words} Rupee")
        else:
            result.append(f"{rupees_words} Rupees")
    else:
        result.append("Zero Rupees")
    
    if paise > 0:
        paise_words = _number_to_words_indian(paise)
        if paise == 1:
            result.append(f"and {paise_words} Paisa")
        else:
            result.append(f"and {paise_words} Paise")
    
    result.append("Only")
    
    return " ".join(result)


def _number_to_words_indian(n: int) -> str:
    """
    Convert number to words using Indian numbering system
    (Ones, Tens, Hundreds, Thousands, Lakhs, Crores)
    """
    if n == 0:
        return "Zero"
    
    # Number word mappings
    ones = [
        "", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine",
        "Ten", "Eleven", "Twelve", "Thirteen", "Fourteen", "Fifteen", "Sixteen",
        "Seventeen", "Eighteen", "Nineteen"
    ]
    
    tens = [
        "", "", "Twenty", "Thirty", "Forty", "Fifty", "Sixty", "Seventy", "Eighty", "Ninety"
    ]
    
    def convert_below_thousand(num: int) -> str:
        """Convert numbers below 1000"""
        if num == 0:
            return ""
        elif num < 20:
            return ones[num]
        elif num < 100:
            return tens[num // 10] + (" " + ones[num % 10] if num % 10 != 0 else "")
        else:
            return (
                ones[num // 100] + " Hundred" +
                (" " + convert_below_thousand(num % 100) if num % 100 != 0 else "")
            )
    
    # Indian numbering system: Crores, Lakhs, Thousands, Hundreds
    if n >= 10000000:  # Crores
        crores = n // 10000000
        remainder = n % 10000000
        result = convert_below_thousand(crores) + " Crore"
        if remainder > 0:
            result += " " + _number_to_words_indian(remainder)
        return result
    
    elif n >= 100000:  # Lakhs
        lakhs = n // 100000
        remainder = n % 100000
        result = convert_below_thousand(lakhs) + " Lakh"
        if remainder > 0:
            result += " " + _number_to_words_indian(remainder)
        return result
    
    elif n >= 1000:  # Thousands
        thousands = n // 1000
        remainder = n % 1000
        result = convert_below_thousand(thousands) + " Thousand"
        if remainder > 0:
            result += " " + convert_below_thousand(remainder)
        return result
    
    else:
        return convert_below_thousand(n)


# Example usage and tests
if __name__ == "__main__":
    test_amounts = [
        Decimal("0.00"),
        Decimal("1.00"),
        Decimal("15.50"),
        Decimal("100.99"),
        Decimal("1234.56"),
        Decimal("12345.67"),
        Decimal("123456.78"),
        Decimal("1234567.89"),
        Decimal("12345678.90"),
    ]
    
    for amount in test_amounts:
        print(f"{amount}: {amount_to_words(amount)}")
